crates/next-api/src/nft_json.rs

```diff
@@ -0,0 +1,191 @@
+};
+
+/// A json file that produces references to all files that are needed by the given module
+/// at runtime. This will include, for example, node native modules, unanalyzable packages,
+/// client side chunks, etc.
+///
+/// With this file, users can determine the minimum set of files that are needed alongside
+/// their bundle.
+#[turbo_tasks::value(shared)]
+pub struct NftJsonAsset {
+    /// The chunk for which the asset is being generated
+    chunk: Vc<Box<dyn OutputAsset>>,
+    output_fs: Vc<DiskFileSystem>,
+    project_fs: Vc<DiskFileSystem>,
+    client_fs: Vc<Box<dyn FileSystem>>,
+    /// Additional assets to include in the nft json. This can be used to manually collect assets
+    /// that are known to be required but are not in the graph yet, for whatever reason.
+    ///
+    /// An example of this is the two-phase approach used by the `ClientReferenceManifest` in
+    /// next.js.
+    additional_assets: Vec<ResolvedVc<Box<dyn OutputAsset>>>,
+}
+
+#[turbo_tasks::value_impl]
+impl NftJsonAsset {
+    #[turbo_tasks::function]
+    pub fn new(
+        chunk: Vc<Box<dyn OutputAsset>>,
+        output_fs: Vc<DiskFileSystem>,
+        project_fs: Vc<DiskFileSystem>,
+        client_fs: Vc<Box<dyn FileSystem>>,
+        additional_assets: Vec<ResolvedVc<Box<dyn OutputAsset>>>,
+    ) -> Vc<Self> {
+        NftJsonAsset {
+            chunk,
+            output_fs,
+            project_fs,
+            client_fs,
+            additional_assets,
+        }
+        .cell()
+    }
+}
+
+#[turbo_tasks::value(transparent)]
+pub struct OutputSpecifier(Option<RcStr>);
+
+#[turbo_tasks::value_impl]
+impl NftJsonAsset {
+    #[turbo_tasks::function]
+    async fn ident_in_project_fs(self: Vc<Self>) -> Result<Vc<FileSystemPath>> {
+        let this = self.await?;
+        let project_fs = this.project_fs.await?;
+        let output_fs = this.output_fs.await?;
+        let nft_folder = self.ident().path().parent().await?;
+
+        if let Some(subdir) = output_fs.root.strip_prefix(&*project_fs.root) {
+            Ok(this
+                .project_fs
+                .root()
+                .join(subdir.into())
+                .join(nft_folder.path.clone()))
+        } else {
+            // TODO: what are the implications of this?
+            bail!("output fs not inside project fs");
+        }
+    }
+
+    #[turbo_tasks::function]
+    async fn ident_in_client_fs(self: Vc<Self>) -> Result<Vc<FileSystemPath>> {
+        Ok(self
+            .await?
+            .client_fs
+            .root()
+            .join(self.ident().path().parent().await?.path.clone()))
+    }
+
+    #[turbo_tasks::function]
+    async fn get_output_specifier(
+        self: Vc<Self>,
+        path: Vc<FileSystemPath>,
+    ) -> Result<Vc<OutputSpecifier>> {
+        let this = self.await?;
+        let path_fs = path.fs().resolve().await?;
+        let path_ref = path.await?;
+        let nft_folder = self.ident().path().parent().await?;
```

Comment:
I think it would be helpful to have a helper method nft_folder() in NftJsonAsset to cache that